### PR TITLE
Allow users to change the base URI for requests

### LINKED ISSUE #58: Should allow baseUrl to be changed for all client objects

### PR DESCRIPTION
**Pull request by ChrisGuzman**
Allow users to change the base URI for requests. Fixes #58

## Contribution Checklist
* [x] Unit tests!
* [x] Updated [CHANGELOG.md](CHANGELOG.md)
* [x] My name is in [CONTRIBUTORS.md](CONTRIBUTORS.md)

### CODE REVIEW

**Comment by ChrisGuzman** [comment]
Do we want a trailing slash on a url or not?

https://github.com/Nexmo/nexmo-java/blob/master/src/main/java/com/nexmo/client/voice/endpoints/CreateCallMethod.java#L42
vs
https://github.com/Nexmo/nexmo-java/blob/master/src/main/java/com/nexmo/client/voice/endpoints/ReadCallMethod.java#L39

**Comment by ChrisGuzman** [comment]
Let's test this IRL

**Comment by ChrisGuzman** [comment]
Hi @stefanscheidt we noticed you thumbs up'ed #58, the issue that this PR aims to resolve. Would you like to look over this PR and let us know your thoughts?

**Comment by ChrisGuzman** [comment]
@judy2k it looks like you had the idea a while back about changing the baseUri for a request https://github.com/Nexmo/nexmo-java/blob/master/src/main/java/com/nexmo/client/voice/endpoints/CreateCallMethod.java#L70

https://github.com/Nexmo/nexmo-java/blob/master/src/test/java/com/nexmo/client/voice/endpoints/CreateCallMethodTest.java#L92

File: src/main/java/com/nexmo/client/insight/InsightClient.java [hunk 1]
```diff
@@ -97,4 +105,12 @@ public AdvancedInsightResponse getAdvancedNumberInsight(
             String number, String country, String ipAddress, boolean cnam) throws IOException, NexmoClientException {
         return this.advanced.execute(new AdvancedInsightRequest(number, country, ipAddress, cnam));
     }
+
+    public String getBaseUri() {
+        return baseUri;
+    }
+
+    public void setBaseUri(String baseUri) {
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/insight/InsightClient.java]: Shouldn't this change the baseUri on all the Endpoint objects?

**Comment by judy2k** [comment]
I knew we wanted to be able to change the baseUri when this stuff was originally developed, but didn't have time before the 2.0 release to implement it.

File: src/main/java/com/nexmo/client/voice/endpoints/StopStreamMethod.java [hunk 2]
```diff
@@ -37,14 +37,19 @@
...
     private static final Class[] ALLOWED_AUTH_METHODS = new Class[]{JWTAuthMethod.class};
     private String uri = DEFAULT_URI;
 
     public StopStreamMethod(HttpWrapper httpWrapper) {
         super(httpWrapper);
     }
 
+    public StopStreamMethod(HttpWrapper httpWrapper, String baseUri) {
+        super(httpWrapper);
+        uri = baseUri + "/v1/calls";
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/endpoints/StopStreamMethod.java]: I think my general preference is to *have* the slash at the end, but I'm not sure whether there's one version that's more correct, or even more of a standard internally. So as long as we're consistent I'm happy and I'll go with whatever you're happy with.

Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/voice/endpoints/StopStreamMethod.java]: I decided to remove the slash at the end since that's the way the library had it before we started making changes and the docs all omit a trailing slash https://developer.nexmo.com/api/voice

File: src/main/java/com/nexmo/client/NexmoClient.java [hunk 3]
```diff
@@ -82,4 +95,11 @@ public VoiceClient getVoiceClient() {
         return this.voice;
     }
 
+    public String getBaseUri() {
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/NexmoClient.java]: Done

File: src/main/java/com/nexmo/client/insight/InsightClient.java (hunk 1 above)
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/insight/InsightClient.java]: Good call. Change made and test updated

File: src/main/java/com/nexmo/client/NexmoClient.java [hunk 4]
```diff
@@ -36,8 +36,6 @@
  * Construct an instance of this object with one or more {@link AuthMethod}s (providing all the authentication methods
  * for the APIs you wish to use), and then call {@link #getVoiceClient()} to obtain a client for the Nexmo Voice API.
  * <p>
- * Currently this object only constructs and provides access to {@link VoiceClient}. In the future it will manage
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/NexmoClient.java]: This isn't true anymore right @judy2k

File: src/test/java/com/nexmo/client/insight/InsightClientTest.java [hunk 5]
```diff
@@ -99,4 +101,21 @@ public void testGetAdvancedNumberInsight3() throws Exception {
         verify(client.advanced).execute(eq(new AdvancedInsightRequest("12345", "GB", "123.123.123.123", true)));
     }
 
+    @Test
+    public void testSetBaseUri() throws Exception {
+        assertNull(this.client.getBaseUri());
+        this.client.setBaseUri("https://example.com");
+        assertEquals("https://example.com", this.client.getBaseUri());
+
+        assertEquals("https://example.com/ni/basic/json", this.client.basic.getUri());
```
Comment (ChrisGuzman) [inline src/test/java/com/nexmo/client/insight/InsightClientTest.java]: These tests fail because we're mocking the clients. Can you help me out with this @judy2k?

Comment (ChrisGuzman) [inline src/test/java/com/nexmo/client/insight/InsightClientTest.java]: Nevermind. Just adjusted the test a bit.

File: src/main/java/com/nexmo/client/insight/basic/BasicInsightEndpoint.java [hunk 6]
```diff
@@ -45,6 +45,20 @@ public BasicInsightEndpoint(HttpWrapper httpWrapper) {
         super(httpWrapper);
     }
 
+    public BasicInsightEndpoint(HttpWrapper httpWrapper, String baseUri) {
+        super(httpWrapper);
+        uri = baseUri + "/ni/basic/json";
```
Comment (stefanscheidt) [inline src/main/java/com/nexmo/client/insight/basic/BasicInsightEndpoint.java]: Probably extract "/ni/basic/json" in a constant like INSIGHT_ENDPOINT_PATH and also use it in setBaseUri?

File: src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpoint.java [hunk 7]
```diff
@@ -44,6 +44,19 @@ public AdvancedInsightEndpoint(HttpWrapper httpWrapper) {
         super(httpWrapper);
     }
 
+    public AdvancedInsightEndpoint(HttpWrapper httpWrapper, String baseUri) {
+        super(httpWrapper);
+        uri = baseUri + "/ni/advanced/json";
```
Comment (stefanscheidt) [inline src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpoint.java]: Probably extract "/ni/advanced/json" in a constant like ADVANCED_ENDPOINT_PATH and also use it in setBaseUri?

File: src/main/java/com/nexmo/client/sms/SendMessageEndpoint.java [hunk 8]
```diff
@@ -57,6 +57,15 @@ public SendMessageEndpoint(HttpWrapper httpWrapper) {
...
 
+    public SendMessageEndpoint(HttpWrapper httpWrapper, String baseUri) {
+        super(httpWrapper);
+        uri = baseUri + "/sms/xml";
+    }
+
+    public String getUri() {
+        return uri;
+    }
+
```
Comment (stefanscheidt) [inline src/main/java/com/nexmo/client/sms/SendMessageEndpoint.java]: No setBaseUri here?

File: src/main/java/com/nexmo/client/sms/SmsClient.java [hunk 9]
```diff
@@ -65,4 +71,11 @@ public SmsClient(HttpWrapper httpWrapper) {
...
     }
 
-}
+     public String getBaseUri() {
+         return baseUri;
+     }
+
+     public void setBaseUri(String baseUri) {
+         this.baseUri = baseUri;
+     }
```
Comment (stefanscheidt) [inline src/main/java/com/nexmo/client/sms/SmsClient.java]: Should a this.message.setBaseUri(baseUri) be called here?

File: src/main/java/com/nexmo/client/sns/SnsClient.java [hunk 10]
```diff
@@ -48,11 +49,25 @@ public SnsClient(HttpWrapper httpWrapper) {
...
         return (SnsSubscribeResponse) this.endpoint.execute(request);
     }
+
+    public String getBaseUri() {
+        return baseUri;
+    }
+
+    public void setBaseUri(String baseUri) {
+        this.baseUri = baseUri;
+    }
```
Comment (stefanscheidt) [inline src/main/java/com/nexmo/client/sns/SnsClient.java]: add this.endpoint.setBaseUri(baseUri) ?

File: src/main/java/com/nexmo/client/sns/SnsEndpoint.java [hunk 11]
```diff
@@ -60,6 +60,15 @@ public SnsEndpoint(HttpWrapper httpWrapper) {
...
 
+    public SnsEndpoint(HttpWrapper httpWrapper, String baseUri) {
+        super(httpWrapper);
+        uri = baseUri + "/sns/xml";
+    }
+
+    public String getUri() {
+        return uri;
+    }
+
```
Comment (stefanscheidt) [inline src/main/java/com/nexmo/client/sns/SnsEndpoint.java]: add setBaseUri?

File: src/test/java/com/nexmo/client/NexmoClientTest.java [hunk 12]
```diff
@@ -76,4 +76,49 @@ public void testConstructNexmoClient() throws Exception {
...
+                "  \"direction\": \"outbound\"\n" +
+                "}"));
+        assertEquals("https://example.com", client.getBaseUri());
+        CallEvent evt = client.getVoiceClient().createCall(new Call("4499991111", "44111222333", "https://callback.example.com/"));
+        assertEquals(CallStatus.STARTED, evt.getStatus());
+    }
+
+
+    @Test
+    public void testBseUriSetterOnNexmoClient() throws Exception {
```
Comment (stefanscheidt) [inline src/test/java/com/nexmo/client/NexmoClientTest.java]: Typo: testBaseUri...

File: src/main/java/com/nexmo/client/sms/SmsClient.java (hunk 9 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/sms/SmsClient.java]: Yeah, @ChrisGuzman, you need to go through each of the XxxClient classes and ensure that `setBaseUri` propagates its value to the Endpoints contained within each client.

File: src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpoint.java (hunk 7 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpoint.java]: I agree with @stefanscheidt . Sorry for setting a bad precedent here, @ChrisGuzman.

I'm happy for this to be done in a separate PR, given that we'd really like to get this released soon.

**Comment by ChrisGuzman** [comment]
@judy2k wanna give this another review?

File: src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpoint.java [hunk 13]
```diff
@@ -36,14 +36,28 @@
...
     private static final Class[] ALLOWED_AUTH_METHODS = new Class[]{SignatureAuthMethod.class, TokenAuthMethod.class};
 
-    private static final String DEFAULT_URI = "https://api.nexmo.com/ni/advanced/json";
+    private static final String DEFAULT_URI  = "https://api.nexmo.com/ni/advanced/json";
+    private static final String DEFAULT_PATH = "/ni/advanced/json";
 
     private String uri = DEFAULT_URI;
 
     public AdvancedInsightEndpoint(HttpWrapper httpWrapper) {
         super(httpWrapper);
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpoint.java]: I always think it's a good idea to funnel all constructor calls through a single "master" constructor. That way, if you want to change the way a class is constructed there's only one place you need to make the change, in theory.

Given this, what do you think of changing `super(httpWrapper)` here to `this(httpWrapper, DEFAULT_BASE_URI)`, and removing DEFAULT_URI?

File: src/main/java/com/nexmo/client/sms/SmsClient.java [hunk 14]
```diff
@@ -65,4 +71,12 @@ public SmsClient(HttpWrapper httpWrapper) {
         return this.message.execute(message);
     }
 
-}
+     public String getBaseUri() {
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/sms/SmsClient.java]: If I construct this with `SmsClient(httpWrapper)` and then call `getBaseUri()`, instead of returning the default, it's going to return `null` and that bugs me a little bit.

File: src/main/java/com/nexmo/client/sns/SnsEndpoint.java [hunk 15]
```diff
@@ -60,6 +62,19 @@ public SnsEndpoint(HttpWrapper httpWrapper) {
...
+    public SnsEndpoint(HttpWrapper httpWrapper, String baseUri) {
+        super(httpWrapper);
+        uri = baseUri + DEFAULT_PATH;
+    }
+
+    public String getUri() {
+        return uri;
+    }
+
+    public void setUri(String uri) {
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/sns/SnsEndpoint.java]: This should really be `setBaseUri()`

File: src/main/java/com/nexmo/client/sms/SmsClient.java (hunk 14 above)
Comment (cr0wst) [reply src/main/java/com/nexmo/client/sms/SmsClient.java]: I know this is an old comment but I think that can be solved similar to your comment about constructors calling through a single master constructor.  The `SmsClient(HttpWrapper)` constructor should call the `SmsClient(HttpWrapper, string)` constructor with the default URI.

File: src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpoint.java (hunk 13 above)
Comment (cr0wst) [reply src/main/java/com/nexmo/client/insight/advanced/AdvancedInsightEndpoint.java]: Again, old comment, but I wanted to chime in that I agree with this.  This should, ideally, also be the common practice for overloaded methods.  Starting with the lowest number of arguments and working your way in.  This also ensures that the object is constructed in a consistent way.

**Comment by cr0wst** [comment]
This PR is a bit older so these comment may not be as relevant:

1. It looks like `baseUri` is becoming a common property between each of the endpoint clients.  Since each client doesn't currently have a parent class, this could be functionality that an abstract class provides through inheritance.

2. I also think it's interesting that the uri passed from the `NexmoClient` to the individual clients and then onto the endpoint.  It's stored on the `NexmoClient`, the individual clients, and eventually stored on the endpoints as part of `uri`.  It seems like there are a lot of places for this URI to be referenced.  Providing statefulness to so many locations doesn't seem completely necessary.  If I were to call `NexmoClient.setBaseUri` what is the gain from storing it on the client, or the individual clients, if it's just going to propagate and override the uri on the individual endpoints?

I also see this as a property of each endpoint, which would make it a good candidate (in my opinion) for something that is in `AbstractMethod` which should maybe just be called `Endpoint` (although I see that sometimes it's extended by a `*Method` and I haven't figured out the difference yet.

**Comment by ChrisGuzman** [comment]
I'm in favor of a `AbstractMethod` approach.

**Comment by judy2k** [comment]
Hey @ChrisGuzman - this PR has decayed, and I think a better approach is to change the Endpoint/Method instances in-place, rather than re-creating when the value changes.  I almost have a solution in #149 

Do you mind if we close this? 😬
